invenio_rdm_migrator/load/postgresql/table_generators/base.py

```diff
@@ -0,0 +1,153 @@
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Base table generator."""
+
+import csv
+import json
+from abc import ABC, abstractmethod
+from dataclasses import fields
+from datetime import datetime
+from uuid import UUID
+
+from ....logging import Logger
+from ....utils import JSONEncoder
+
+
+# PORT: from invenio_records.dictutils to avoid having an invenio constraint
+# it could cause troubles with sqlalchemy and psycompg version
+def parse_lookup_key(lookup_key):
+    """Parse a lookup key."""
+    if not lookup_key:
+        raise KeyError("No lookup key specified")
+
+    # Parse the list of keys
+    if isinstance(lookup_key, str):
+        keys = lookup_key.split(".")
+    elif isinstance(lookup_key, list):
+        keys = lookup_key
+    else:
+        raise TypeError("lookup must be string or list")
+
+    return keys
+
+
+def dict_set(source, key, value):
+    """Set a value into a dict via a dot-notated key."""
+    keys = parse_lookup_key(key)
+    parent = source
+    for key in keys[:-1]:
+        if isinstance(key, int):
+            parent = parent[key]
+        else:
+            parent = parent.setdefault(key, {})
+    parent[keys[-1]] = value
+
+
+def as_csv_row(dc):
+    """Serialize a dataclass instance as a CSV-writable row."""
+    row = []
+    for f in fields(dc):
+        val = getattr(dc, f.name)
+        if val:
+            if issubclass(f.type, (dict,)):
+                val = json.dumps(val, cls=JSONEncoder)
+            elif issubclass(f.type, (datetime,)):
+                val = val.isoformat()
+            elif issubclass(f.type, (UUID,)):
+                val = str(val)
+        row.append(val)
+    return row
+
+
+class TableGenerator(ABC):
+    """Table generator interface."""
+
+    def __init__(self, pks=None, post_load_hooks=None):
+        """Constructor."""
+        self.post_load_hooks = post_load_hooks or []
+        self.pks = pks or []
+
+    def _generate_rows(self, **kwargs):
+        """Yield generated rows."""
+        # raises an error but does not force an implementation
+        # e.g. when `prepare` is overwritten, _generate_rows is not required
+        raise NotImplementedError
+
+    def cleanup(self, **kwargs):
+        """Cleanup."""
+        pass
+
+    def _generate_pks(self, data, create=False):
+        keys = data.keys()
+        for path, pk_func in self.pks:
+            try:
+                root = path.split(".")[0]
+                # avoids creating e.g. "record" in a draft and generating a recid + uuid
+                if create or root in keys:
+                    dict_set(data, path, pk_func(data))
+            except KeyError:
+                logger = Logger.get_logger()
+                logger.exception(f"Path {path} not found on record")
+
+    def _resolve_references(self, data, **kwargs):
+        """Resolve references e.g communities slug names."""
+        pass
+
+    @abstractmethod
+    def prepare(self, **kwargs):
```

Comment:
Each (or almost each) stream will need a CSV table generator for bulk/initial migration, and then another in "memory" one for the action streaming. This second one, will most likely be "ligther" and yield only a part what the CSV does, since it's not grouped (e.g. PIDs are created by a different TG than the records).

In the current implementation, this ends up in very nested subclassing and incompatible APIs (both in arguments and return types, rendering polymorphism null). I am considering the option to split preparation from row generation (e.g. mixins or function composition).